ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDFConvertCharset.java

```diff
@@ -0,0 +1,126 @@
+    }
+
+    checkInputArgument(arguments, 0);
+    stringOI = (PrimitiveObjectInspector) arguments[0];
+
+    checkInputArgument(arguments, 1);
+    fromCharsetOI = (PrimitiveObjectInspector) arguments[1];
+
+    checkInputArgument(arguments, 2);
+    toCharsetOI = (PrimitiveObjectInspector) arguments[2];
+
+    // If the character set for encoding is constant, we can optimize that
+    if (fromCharsetOI instanceof ConstantObjectInspector) {
+      String charSetName = ((ConstantObjectInspector) arguments[1]).getWritableConstantValue().toString();
+      encoder = Charset.forName(charSetName).newEncoder().onMalformedInput(CodingErrorAction.REPORT)
+          .onUnmappableCharacter(CodingErrorAction.REPORT);
+    }
+
+    // If the character set for decoding is constant, we can optimize that
+    if (toCharsetOI instanceof ConstantObjectInspector) {
+      String charSetName = ((ConstantObjectInspector) arguments[2]).getWritableConstantValue().toString();
+      decoder = Charset.forName(charSetName).newDecoder().onMalformedInput(CodingErrorAction.REPORT)
+          .onUnmappableCharacter(CodingErrorAction.REPORT);
+    }
+
+    return PrimitiveObjectInspectorFactory.javaStringObjectInspector;
+  }
+
+  @Override public Object evaluate(DeferredObject[] arguments) throws HiveException {
+    String value = PrimitiveObjectInspectorUtils.getString(arguments[0].get(), stringOI);
+    if (value == null) {
+      return null;
+    }
+
+    ByteBuffer encoded;
+    if (encoder != null) {
+      try {
+        encoded = encoder.encode(CharBuffer.wrap(value));
+      } catch (CharacterCodingException e) {
+        throw new HiveException(e);
+      }
+    } else {
+      encoded =
+          Charset.forName(PrimitiveObjectInspectorUtils.getString(arguments[1].get(), fromCharsetOI)).encode(value);
+    }
+
+    CharBuffer decoded;
+    if (decoder != null) {
+      try {
+        decoded = decoder.decode(encoded);
+      } catch (CharacterCodingException e) {
+        throw new HiveException(e);
+      }
+    } else {
+      decoded =
+          Charset.forName(PrimitiveObjectInspectorUtils.getString(arguments[2].get(), toCharsetOI)).decode(encoded);
+    }
+    return decoded.toString();
+  }
+
+  @Override public String getDisplayString(String[] children) {
+    assert (children.length == 3);
+    return getStandardDisplayString("convertCharset", children, ",");
+  }
+
+  private void checkInputArgument(ObjectInspector[] arguments, int index) throws UDFArgumentTypeException {
+    if (arguments[index].getCategory() != ObjectInspector.Category.PRIMITIVE
+        || PrimitiveObjectInspectorUtils.PrimitiveGrouping.STRING_GROUP
+        != PrimitiveObjectInspectorUtils.getPrimitiveGrouping(
+        ((PrimitiveObjectInspector) arguments[index]).getPrimitiveCategory())) {
+      throw new UDFArgumentTypeException(index, "The argument to ConvertCharset() must be a string/varchar");
```

Comment:
`PrimitiveGrouping.STRING_GROUP` also contains char datatype in its group. 
https://github.com/apache/hive/blob/master/serde/src/java/org/apache/hadoop/hive/serde2/objectinspector/primitive/PrimitiveObjectInspectorUtils.java#L1434-L1437

Please reflect the same in the exception thrown.